### ISSUE TITLE
Loading multiple variants of same base image?

### ISSUE THREAD

**Issue by dlindenkreuz** [kean/Nuke#182]
I am trying to load one image in separate variants: (a) resized, (b) resized + grayscale, (c) resized + grayscale + blur.

My current implementation has Nuke load the images three times in parallel and does some processing as described in the [docs](https://github.com/kean/Nuke#process-an-image). This way, it seems the original image gets read from disk and decompressed three times, which is pretty inefficient of course.

Instead, I'd like to load the image from disk once, create + cache variant (a), create + cache variant (b) based on the results of (a), and then create + cache variant (c) based on the results of (b). Is there an elegant way to do this?

**Comment by kean** [comment]
Hi, there is no elegant way to do this at the moment.

**Comment by swftvsn** [comment]
I'd like this feature too. Looking at the roadmap, there's something coming related to thumbnail caching. Maybe this could be altered a bit to include multiple versions cached, produced by a set (or multiple variants using multiple sets) of processors? Ie. provide multiple sets of processors and indicate which set to use for the current load, then use the rest to cache on disk.

Anyhow, you can catch the raw data from the load and manipulate + fill caches as necessary by piggybacking the URLCache:

```swift
class InterceptingURLCache: URLCache {
    override open func storeCachedResponse(_ cachedResponse: CachedURLResponse, for request: URLRequest) {
        NukeImageService.instance.cacheData(url: request.url!.absoluteString, data: cachedResponse.data)
    }
    override open func storeCachedResponse(_ cachedResponse: CachedURLResponse, for dataTask: URLSessionDataTask) {
        NukeImageService.instance.cacheData(url: dataTask.originalRequest!.url!.absoluteString, data: cachedResponse.data)
    }
}
```

and then in url session config:
```swift
conf.urlCache = InterceptingURLCache(memoryCapacity: 1024*1024*1024, diskCapacity: 0, diskPath: nil)
```

Note to set a large memoryCapacity, as otherwise the url cache is not called at all. The example actually does not store anything to underlying cache, so no memory is consumed.

Not pretty, but I didn't find any other way :(

**Comment by kean** [comment]
Yeah, this is one way to achieve this.

> Looking at the roadmap, there's something coming related to thumbnail caching. 

There is lots of stuff in the backlog, I wouldn't hold my breath on these features though, I'm super busy at work right now :(

**Comment by kean** [comment]
Hi, @dlindenkreuz, are you still interested in this feature?

I'm currently designing an API to support a related but simpler feature: "As a user, I would like to store processed images on disk instead of the original ones, because recreating the processed ones is expensive and storing the original ones is wasteful".

With this feature in place it would also be possible to allow users to store both original and processed images which could be a (partial) solution for your use case – you would be able to load the original image data once, and then process the images (on demand) and store the processed ones. But it won't allow you to pass images between processors.

Getting rid of the duplicated work when processing images is also possible by caching intermediate results – I already designed a system like that once.

**Comment by dlindenkreuz** [comment]
Yup, I still need this in one way or another 😉 This would already alleviate resource usage in most of my cases. Thanks for getting back to this and letting me know!

**Comment by kean** [comment]
Oh, that’s cool. It’s an interesting challenge. I’m currently working on describing the technical solution in detail and will hopefully post it some time soon.

**Comment by kean** [comment]
So, there are at least five things that need to happen to make this all work:

- [#227 Caching Processed Image](https://github.com/kean/Nuke/pull/227)
    - [#228 ImageProcessing String identifiers](https://github.com/kean/Nuke/pull/228)
    - [#229  Decoupling Image Decompression from Processing](https://github.com/kean/Nuke/pull/229)
    - Image Encoders (no PR yet)
- Reusing Intermediate Processing Results (no PR yet)

So far, I'm very happy with most of this change, especially with https://github.com/kean/Nuke/pull/229 because it finally allowed me to rework decompression and hopefully finally get it right. I think it's now both more efficient than in other frameworks _and_ just as easy to use.

**Comment by kean** [comment]
Hey, @dlindenkreuz. There are two new features available in [Nuke 8 betas](https://github.com/kean/Nuke/releases) which address your use case:

- [#227 Caching Processed Images](https://github.com/kean/Nuke/pull/227)
- [#247 Deduplicating Processing Tasks](https://github.com/kean/Nuke/pull/247)

If have some time to spare, please give it a try.

**Comment by kean** [comment]
Shipped in [Nuke 8.0-rc.1](https://github.com/kean/Nuke/releases/tag/8.0-rc.1)
